mindsdb/integrations/handlers/anyscale_endpoints_handler/anyscale_endpoints_handler.py

```diff
@@ -119,63 +120,35 @@ def _check_ft_cols(df, cols):
             if col not in set(df.columns):
                 raise Exception(f"To fine-tune this model, format your select data query to have a `role` column and a `content` column.")  # noqa
 
-    def _prepare_ft_jsonl(self, df, temp_storage_path, temp_filename, _, test_size=0.2):
-        """
-            df: has exactly two columns, `role` and `content`. Rows contain >= 1 chats in long (stacked) format.
-            For more details, check `FineTuning -> Data Format` in the Anyscale API reference.
-        """
-        def _is_valid(chat):
-            """ Check if chat is valid according to Anyscale criteria."""
-            roles = [m['role'] for m in chat]
-            transitions = {None: ['system', 'user'], 'system': ['user'], 'user': ['assistant'], 'assistant': ['user']}
-
-            # check base condition
-            if not ('user' in roles and 'assistant' in roles):
-                return False
-
-            # check order is valid
-            state = None
-            for role in roles:
-                if role not in transitions[state]:
-                    return False
-                else:
-                    state = role
-
-            # chat is valid, return
-            return True
-
-        # 1. aggregate each chat sequence into one row
-        chats = []
-        chat = []
-        for i, row in df.iterrows():
-            if row['role'] == 'system' and len(chat) > 0:
-                if _is_valid(chat):
-                    chats.append({'messages': chat})
-                chat = []
-            event = {'role': row['role'], 'content': row['content']}
-            chat.append(event)
-
-        if _is_valid(chat):
-            chats.append({'messages': chat})
+    # TODO: add unit tests
+    @staticmethod
+    def _prepare_ft_jsonl(df, temp_storage_path, temp_filename, _, test_size=0.2):
+        # 1. format data
+        chats = ft_chat_formatter(df)
 
+        # 2. split chats in training and validation subsets
         series = pd.Series(chats)
         if len(series) < 20 * 2:
```

Comment:
This seems a little arbitrary; shouldn't the length of the dataset depend on how large the model is? I would at least leave that "20" as a keyword parameter with a default value.

Reply:
Good shout, I'll change the row number to be customizable. As for why: Anyscale endpoints has a non-documented cutoff at 20 -- if the validation dataset is smaller than that, the entire process fails. I'll add this info. as a comment.

Could you expand on how I should go about this one?
> I strongly suggest to refactor the code a little bit.